ramips: mt7621: refactor set affinity script

dengqf6 opened the pull request:
The current one only looks for mt76x2e and mt7603e, and
does not work for 2 or more same Wi-Fi chips.
Refactor the script to cover those cases.

ping @nbd168

xabolcs commented:
Hello!

If you already rewrote this script and changing behavior, it would be nice to assign the mt76 interrupts to physical cores equally:
* if the SoC is 2C/4T: then the assigning function should loop through the CPU0 -> CPU2 -> CPU1 -> CPU3 list
* if the SOC is 1C/2T: then the assigning function should loop through the CPU0 -> CPU1 list

I'm unsure how much mt76 interrupts can it be in an OpenWrt supported router. :sweat_smile: 

My DIR-860L is affected by your proposal: it has two `mt76x2e` interrupts and has 2 cores with 4 threads. :)
```bash
root@OpenWrt:~# grep "mt76..e" /proc/interrupts | cut -d: -f1 | sed 's, *,,'                                            
23                                                                                                                      
24 
```
```bash
root@OpenWrt:~# grep -e processor -e core -i -e vpe /proc/cpuinfo                                                       
processor               : 0                                                                                             
core                    : 0                                                                                             
VPE                     : 0                                                                                             
processor               : 1                                                                                             
core                    : 0                                                                                             
VPE                     : 1                                                                                             
processor               : 2                                                                                             
core                    : 1                                                                                             
VPE                     : 0                                                                                             
processor               : 3                                                                                             
core                    : 1                                                                                             
VPE                     : 1                                                                                             
```

 
The original script didn't touch the second, so it was allowed on all threads:
```bash
root@OpenWrt:~# cat /proc/irq/24/smp_affinity_list                                                                      
0-3
```

Your change assigns the second mt76 interrupt to CPU3 which is the same physical core as CPU2 where to the first mt76 is assigned.

```bash
root@OpenWrt:~# grep "mt76..e" /proc/interrupts                                                                         
 23:          2          0      23701          0  MIPS GIC  11  mt76x2e                                                 
 24:      73812          0          0          0  MIPS GIC  31  mt76x2e                                                 
root@OpenWrt:~# echo 8 > /proc/irq/24/smp_affinity                                                                      
root@OpenWrt:~# cat /proc/irq/24/smp_affinity_list                                                                      
3                                                                                                                       
root@OpenWrt:~# grep "mt76..e" /proc/interrupts                                                                         
 23:          2          0      24863          0  MIPS GIC  11  mt76x2e                                                 
 24:      77871          0          0       1719  MIPS GIC  31  mt76x2e                                                 
root@OpenWrt:~# cat /proc/irq/23/smp_affinity_list                                                                      
2                                                                                                                       
```

dengqf6 commented:
@xabolcs 
The script is to **move mt76 interrupts off the main core** (see commit 01895c315f0485779a2e312c58708a32b4ea1f9a), not balance them
The common use case is WAN/LAN to Wi-Fi, where the first core handles eth0's interrupt and the second handles Wi-Fi's

neheb commented on target/linux/ramips/mt7621/base-files/etc/init.d/set-irq-affinity:
```diff
@@ -2,30 +2,20 @@
+	for irq in $(grep "mt76..e" /proc/interrupts | cut -d: -f1 | sed 's, *,,')
+	do
+		echo "$mask" > "/proc/irq/$irq/smp_affinity"
+		if [ $mask == 4 ]; then
```
== is a bashism.

dengqf6 replied:
So I should use `=`?

adrianschmutzler replied:
Yes, please.

adrianschmutzler commented on target/linux/ramips/mt7621/base-files/etc/init.d/set-irq-affinity:
```diff
@@ -2,30 +2,20 @@
 
-	set_irq_affinity mt76x2e $mask1
-	set_irq_affinity mt7603e $mask2
+	for irq in $(grep "mt76..e" /proc/interrupts | cut -d: -f1 | sed 's, *,,')
```
I don't really understand what this is about, but from the code it looks to me like order of interfaces can change here.
In the "old" version, mt76x2 always got mask1 and mt7603e got mask2.

In the for loop, the order of returned strings would be inverted if sorted, and should be random in the current "new" implementation. However, I don't know whether this is a problem for the actual use case.

dengqf6 replied:
So all `mt76xxe` IRQs are moved to the 2nd CPU. Order does not matter.

adrianschmutzler commented on target/linux/ramips/mt7621/base-files/etc/init.d/set-irq-affinity:
```diff
@@ -2,30 +2,20 @@
+		echo "$mask" > "/proc/irq/$irq/smp_affinity"
+		if [ $mask = 4 ]; then
+			mask=8
+		fi
```
Might be shortened into
`[ $mask = 4 ] && mask=8`
But purely optional, as it's matter of taste.

Commit:
ramips: mt7621: refactor set affinity script

The current one only looks for mt76x2e and mt7603e, and
does not work for 2 or more same Wi-Fi chips.
Refactor the script to cover those cases.

Signed-off-by: DENG Qingfang <dengqf6@mail2.sysu.edu.cn>

dengqf6 commented:
@adrianschmutzler ping

adrianschmutzler commented:
@ynezz This has been merged in 23de93a5216bf1c24c4e4c71f6cad4455c3cfa2c
Any reason for keeping it open?
